[CASSANDRA-20167][trunk] Reduce memory allocations in miscellaneous places along write path

netudima opened the pull request:
Avoid unnecessary allocations in different places along write path
- avoid iterator allocations if possible
- handle typical cases (such as a single row, single table writes) more efficiently
- add fast paths for typical scenarios (like pending ranges is empty)
- memorize things which can be computed once

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

Commit:
Reduce memory allocation for write hot path.
Avoid objects allocations used only in rarely invoked logic branches. Memorize things which can be computed once. Avoid unnecessary iterators for the write hot path. Do not recreate immutable objects with the same content.

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

Commit:
Reduce memory allocation in CQL transport logic:
- parse header and query options flags as int, without EnumSet overheads
- avoid ByteBuffer slicing in 2 places
- avoid a capturing lambda allocation by using an explicit interface

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

Commit:
Reduce memory allocation Mutation logic:
- avoid an unnecessary byte buffer slicing (DataOutputBuffer.toByteArray()) during a Mutation serialization
- replace AtomicLong viewLockAcquireStart field in Mutation with volatile long and AtomicLongFieldUpdater usage

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

Commit:
Use simplified iterator logic for PartitionUpdate to avoid LeafBTreeSearchIterator within inclusionTester() logic which checks if we need to iterate over particular column for each cell. It is not really needed here: we've just added these cells during PartitionUpdate creation

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

Commit:
Cache BtreeRow.builder in a thread-local tiny pool to reuse in UpdateParameters

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

Commit:
Add optimized version of UpdatesCollector for a single partition update within a single table ( very typical case when we just modify a single row)

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

Commit:
Avoid ImmutableList.copyOf(columnSpecs) within OptionsWithColumnSpecifications constructor, we had to copy columnSpecs for every request because in one place plain List is used and ImmutableList in another one

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever for CASSANDRA-20167

vlsi commented on src/java/org/apache/cassandra/db/Keyspace.java:
```diff
@@ -534,7 +534,7 @@ else if (isDeferrable)
             }
 
-            long acquireTime = currentTimeMillis() - mutation.viewLockAcquireStart.get();
+            long acquireTime = currentTimeMillis() - mutation.viewLockAcquireStart;
```
Frankly, it sounds a bit odd to have both `plain` and `AtomicLongFieldUpdater`-driven access at the same time. Does it really cost much to have plain `mutation` read rather than `Mutation.viewLockAcquireStartUpdater.get(mutation)` instead?

In theory, the underlying platform might fail to support atomic CAS for long, so OpenJDK would resort to `synchronized...` for `compareAndSet`, thus it would be somewhat invalid to use plain reads in those cases.

netudima replied:
the idea of this change is to reduce allocation by remove AtomicLong in Mutation object (which we create for each request) and replace it with a volatile long + Updater usage

netudima replied:
agree, there is a cost: https://shipilev.net/blog/2015/faster-atomic-fu/
but for the current scenario it is not that critical (especially after Java 9 improvements) - usually we do the operation only once per mutation.

Commit:
Use a more correct way to read a volatile field if AtomicLongFieldUpdater is used

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink, Michael Semb Wever, Vladimir Sitnikov for CASSANDRA-20167
